builder/assignmentEnv.go

```diff
@@ -0,0 +1,315 @@
+				found = true
+			}
+		}
+		if !found {
+			return errors.New("code-runner base image not found on docker registry")
+		}
+	}
+
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) writeFromBaseImage() {
+	assgnEnv.DockerfileData.WriteString(assgnEnv.AssgnEnvConfig.WriteInstruction())
+	// Append library names to image tag.
+	for lib := range assgnEnv.AssgnEnvConfig.Deps.Libraries {
+		assgnEnv.ImgBuildConfig.imageTag = strings.Join([]string{assgnEnv.ImgBuildConfig.imageTag, lib}, "-")
+	}
+}
+
+func (assgnEnv *assignmentEnv) writeFromDependencies() {
+	buf := &bytes.Buffer{}
+	from := fmt.Sprintf("FROM %s", assgnEnv.ImgBuildConfig.imageTag)
+	buf.WriteString(from)
+	buf.WriteString("\n")
+
+	for lib, installCmd := range assgnEnv.AssgnEnvConfig.Deps.Libraries {
+		// Append library names to image tag.
+		assgnEnv.ImgBuildConfig.imageTag = strings.Join([]string{assgnEnv.ImgBuildConfig.imageTag, lib}, "-")
+
+		buf.WriteString("RUN " + installCmd.WriteInstruction() + " " + lib)
+		buf.WriteString("\n")
+	}
+
+	assgnEnv.DockerfileData.WriteString(buf.String())
+}
+
+func (assgnEnv *assignmentEnv) writeDockerfile() error {
+	if assgnEnv.DockerfileData.Len() > 0 {
+		file, err := os.Create(assgnEnv.ImgBuildConfig.dockerFilepath)
+		defer func() {
+			err = file.Close()
+			if err != nil {
+				log.Println("error while closing the created Dockerfile", err)
+				return
+			}
+		}()
+		if err != nil {
+			return err
+		}
+		_, err = file.WriteString(assgnEnv.DockerfileData.String())
+		return err
+	}
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) build() error {
+
+	if assgnEnv.DockerfileData.Len() > 0 {
+		backgroundContext := context.Background()
+		dockerClient, err := client.NewEnvClient()
+		if err != nil {
+			return errors.Wrap(err, "error in creating a docker client")
+		}
+
+		// Create a build context tar for the image.
+		// build Context is the current working directory and where the Dockerfile is assumed to be located.
+		// [cite: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/].
+		dockerFilepath := assgnEnv.ImgBuildConfig.dockerFilepath
+
+		dockerBuildContext, err := assgnEnv.ImgBuildConfig.getDockerBuildContextTar()
+		if err != nil {
+			return err
+		}
+
+		response, err := dockerClient.ImageBuild(
+			backgroundContext,
+			dockerBuildContext,
+			types.ImageBuildOptions{
+				Dockerfile: dockerFilepath,
+				Tags:       []string{assgnEnv.ImgBuildConfig.imageTag}})
+		if err != nil {
+			return errors.Wrap(err, "error in building docker image")
+		}
+		defer func() {
+			err = response.Body.Close()
+			if err != nil {
+				log.Println(err)
+				return
+			}
+		}()
+
+		_, err = io.Copy(os.Stdout, response.Body)
+		if err != nil {
+			return errors.Wrap(err, "error in reading image build response")
+		}
+	}
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) publish() error {
+	if assgnEnv.DockerfileData.Len() <= 0 {
+		return assgnEnv.pullImage()
```

Comment:
why is publish() responsible for pulling the image? Also, what is the reason for pulling the image?

Reply:
As said earlier, if user is attempting to build an image that is already present on hub then simply the image will be pulled . However, the name publish() misleads, so I will change it